api/search.rb

```diff
@@ -1,105 +1,152 @@
-  if !search_text
-    {}.to_json
-  else
-    # Because threads and comments are currently separate unrelated documents in
-    # Elasticsearch, we must first query for all matching documents, then
-    # extract the set of thread ids, and then sort the threads by the specified
-    # criteria and paginate. For performance reasons, we currently limit the
-    # number of documents considered (ordered by update recency), which means
-    # that matching threads can be missed if the search terms are very common.
-
-    get_matching_thread_ids = lambda do |search_text|
-      self.class.trace_execution_scoped(["Custom/get_search_threads/es_search"]) do
-        search = Tire.search Content::ES_INDEX_NAME do
-          query do
-            match [:title, :body], search_text, :operator => "AND"
-            filtered do
-              filter :term, :commentable_id => local_params["commentable_id"] if local_params["commentable_id"]
-              filter :terms, :commentable_id => local_params["commentable_ids"].split(",") if local_params["commentable_ids"]
-              filter :term, :course_id => local_params["course_id"] if local_params["course_id"]
-              filter :or, [
-                {:not => {:exists => {:field => :context}}},
-                {:term => {:context => context}}
-              ]
-
-              if not group_ids.empty?
-                if group_ids.length > 1
-                  group_id_criteria = {:terms => {:group_id => group_ids}}
-                else
-                  group_id_criteria = {:term => {:group_id => group_ids[0]}}
-                end
-
-                filter :or, [
-                  {:not => {:exists => {:field => :group_id}}},
-                  group_id_criteria
-                ]
-              end
-
-            end
-          end
-          sort do
-            by "updated_at", "desc"
-          end
-          size CommentService.config["max_deep_search_comment_count"].to_i
-        end
-        thread_ids = Set.new
-        search.results.each do |content|
-          case content.type
-          when "comment_thread"
-            thread_ids.add(content.id)
-          when "comment"
-            thread_ids.add(content.comment_thread_id)
-          end
-        end
-        thread_ids
+def get_thread_ids(context, group_ids, local_params, search_text)
+  filters = []
+  filters.push({term: {commentable_id: local_params['commentable_id']}}) if local_params['commentable_id']
+  filters.push({terms: {commentable_id: local_params['commentable_ids'].split(',')}}) if local_params['commentable_ids']
+  filters.push({term: {course_id: local_params['course_id']}}) if local_params['course_id']
+
+  filters.push({or: [
+      {not: {exists: {field: :context}}},
+      {term: {context: context}}
+  ]})
+
+  unless group_ids.empty?
+    if group_ids.length > 1
+      group_id_criteria = {:terms => {:group_id => group_ids}}
+    else
+      group_id_criteria = {:term => {:group_id => group_ids[0]}}
+    end
```

Comment:
May be I am wrong, but should not `group_id_criteria = {:terms => {:group_id => group_ids}}` work for `group_ids.length == 1` too; why else is required?